components/rule-mgt/org.wso2.carbon.identity.rule.management/src/main/java/org/wso2/carbon/identity/rule/management/util/RuleBuilder.java

```diff
@@ -216,4 +238,91 @@ private void setValidationError(String message) {
+
+        if (!isValidOptionsInputValue(fieldDefinition, rawValue)) {
+            return value;
+        }
+
+        try {
+            return resolveValue(fieldDefinition, rawValue);
+        } catch (RuleManagementClientException e) {
+            setValidationError(e.getMessage());
+            return value;
+        }
+    }
+
+    private Value resolveValue(FieldDefinition fieldDefinition,
+                               String rawValue) throws RuleManagementClientException {
+
+        org.wso2.carbon.identity.rule.metadata.model.Value.ValueType
+                fieldDefinitionValueType = fieldDefinition.getValue().getValueType();
+
+        switch (fieldDefinitionValueType) {
+            case STRING:
+                return new Value(Value.Type.STRING, rawValue);
+            case NUMBER:
+                return validateNumberValue(rawValue);
+            case BOOLEAN:
+                return validateBooleanValue(rawValue);
+            case REFERENCE:
+                return new Value(Value.Type.REFERENCE, rawValue);
+            default:
+                throw new RuleManagementClientException(
+                        "Unsupported value type: " + fieldDefinitionValueType + " for field: " +
+                                fieldDefinition.getField().getName());
+        }
+    }
+
+    private boolean isValidValueType(FieldDefinition fieldDefinition, Value value) {
+
+        org.wso2.carbon.identity.rule.metadata.model.Value.ValueType fieldDefinitionValueType =
+                fieldDefinition.getValue().getValueType();
+        Value.Type valueType = value.getType();
+
+        if (valueType != Value.Type.RAW && !valueType.name().equals(fieldDefinitionValueType.name())) {
+            setValidationError(
+                    "Value type " + valueType + " is not supported for field " + fieldDefinition.getField().getName());
+            return false;
+        }
+        return true;
+    }
+
+    private boolean isValidOptionsInputValue(FieldDefinition fieldDefinition, String fieldValue) {
+
+        if (fieldDefinition.getValue() instanceof OptionsInputValue &&
+                ((OptionsInputValue) fieldDefinition.getValue()).getValues().stream()
+                        .noneMatch(optionsValue -> optionsValue.getName().equals(fieldValue))) {
+            setValidationError(
+                    "Value " + fieldValue + " is not supported for field " + fieldDefinition.getField().getName());
+            return false;
+        }
+
+        return true;
+    }
+
+    private Value validateNumberValue(String rawValue) throws RuleManagementClientException {
+
+        try {
+            Double.parseDouble(rawValue);
+            return new Value(Value.Type.NUMBER, rawValue);
+        } catch (NumberFormatException e) {
+            throw new RuleManagementClientException("Value " + rawValue + " is not a valid NUMBER.");
+        }
+    }
+
+    private Value validateBooleanValue(String rawValue) throws RuleManagementClientException {
```

Comment:
```suggestion
    private Value validateBooleanValue(String rawValue) throws RuleManagementClientException {
```

Shall we rename this to `validateAndResolveBooleanValue` for better readability

Reply:
Let's keep them. I feel it's too much to change those method names and enough readability is there